Multiplex network service

beanpole135 opened the pull request:
Properly multiplex the network service and have it startup any device-specific services properly based on device config.

This fixes the issues we are seeing with dhcpcd/wpa_supplicant not getting started properly on bootup on many laptops, and also ensures that the service can be used to start/stop/restart network devices individually as needed.

beanpole135 committed:
Multiplex network service

Properly multiplex the network service and have it startup any device-specific services properly based on device config.

kmoore134 commented:
This change scares me a bit. Right now we're doing a mostly "FreeBSD" thing where devd handles the starting of dhcpcd / wpa_supplicant on link-up (How FreeBSD rc.d works). This looks like it could cause weird race conditions now since network is going to try and start services and devd is going to also try to start them, possibly at same time...

beanpole135 committed:
Update network

One addition patch:
Make sure that the sub-services only "stop" the individual device, and don't destroy them. Let the main "network" service create/destroy devices themselves.

beanpole135 commented:
The network service update avoids those race conditions by ensuring that it always starts *before* devd. This allows the init on-boot device setup to be handled by the meta-service, and after devd gets started then the individual hotplugged devices can be added/removed via devd events.

beanpole135 committed:
Update dhcpcd

Make sure the dependencies for dhcpcd are set on the network.[device] and optionally wpa_supplicant.[device] (wlan* devices)

beanpole135 committed:
Update wpa_supplicant

Add dependency for wpa_supplicant.[device] on network.[device]

beanpole135 committed:
Update network

Cleanup some whitespace and add a better check/init of extra services in start_post()

kmoore134 commented on etc/init.d/network:
```diff
@@ -13,23 +14,36 @@
 [ -n "$network_device" ] && name="$name ($network_device)"
 
+if [ -n "${network_device}" ] ; then
+  _conf=`sysrc -n ifconfig_${network_device}`
```
This is unnecessary. /etc/rc.conf and related files are sourced before init.d scripts are run. The environment variables should already be set. Check ifconfig_getargs() in network.subr

kmoore134 commented on etc/init.d/network:
```diff
@@ -64,34 +145,49 @@ start()
 {
-	if [ "$RC_CMD" = "restart" ]; then
+  extra_services_stop
+	if [ "$RC_CMD" = "restart" ] || [ -n "${network_device}" ] ; then
```
This has potential to break other things such as hostap... Services that don't want to "destroy" the device shouldn't be using start/stop and instead use "restart" as FreeBSD does. The way this is written now, you can't destroy any specific device without destroying them *all* which would be a problem for FreeNAS / lagg, etc, etc.

kmoore134 commented on etc/init.d/network:
```diff
@@ -45,16 +59,83 @@ depend()
 	keyword -jail -prefix -vserver -stop
 }
 
+
```
I'm still investigating this. It's interaction with devd is going to be a problem going forward. Right now we assume devd is the source of truth when starting interfaces which aren't explicitly enabled, such as adding network.em0 to default runlevel manually.

kmoore134 reviewed (changes requested):
I'll need to perhaps investigate a bit more on the systems failing with the current implementation. This has a bunch of negative interactions with other services which we may not need to do since it worked perfectly fine with rc.d scripts and devd previously.

beanpole135 committed:
Update network.subr

Make sure that dhcpcd gets started in asynchronous mode if specified by the config.

beanpole135 committed:
Update dhcpcd

Move from need to after for the network dependency in dhcpcd

beanpole135 committed:
Update wpa_supplicant

Make sure that the wpa_supplicant service "needs" network

beanpole135 committed:
Update network

Get the network service working with the "new" /etc/network.subr
Parallelize the init of the individual device services, remove the "extra" services starting (same thing in network.subr now)

beanpole135 commented:
ok, after a lot more work with Kris offline, I have a a finished set of changes to fix the issues we kept uncovering.
I am closing this pull request and will open a new PR for a "clean" view/test of this change.
